File: 2. Implement a stack using linked list.py
"""
Task: Implement a stack using linked list

Time Complexity: push/pop/top O(1)
Space Complexity: O(n)
"""

class Node:
    def __init__(self, val: int):
        self.val = val
        self.next = None

class LinkedListStack:
    def __init__(self):
        self.head = None

    def push(self, x: int):
        n = Node(x)
        n.next = self.head
        self.head = n

    def pop(self):
        if not self.head:
            raise IndexError('pop from empty stack')
        val = self.head.val
        self.head = self.head.next
        return val

    def top(self):
        if not self.head:
            raise IndexError('top from empty stack')
        return self.head.val

    def is_empty(self):
        return self.head is None


if __name__ == '__main__':
    s = LinkedListStack()
    s.push(10)
    s.push(20)
    print(s.pop())
    print(s.top())

File: 2. Implement a stack using linked list.cpp
/*
 Task: Implement a stack using linked list
*/

#include <bits/stdc++.h>
using namespace std;

struct Node
{
    int val;
    Node* next;
    explicit Node(int v) : val(v), next(nullptr) {}
};

class LinkedListStack
{
   public:
    LinkedListStack() : head(nullptr) {}
    ~LinkedListStack()
    {
        while (head)
        {
            Node* t = head;
            head = head->next;
            delete t;
        }
    }

    void push(int x)
    {
        Node* n = new Node(x);
        n->next = head;
        head = n;
    }
    bool pop(int& out)
    {
        if (!head)
            return false;
        out = head->val;
        Node* t = head;
        head = head->next;
        delete t;
        return true;
    }
    bool isEmpty() const { return head == nullptr; }
    int top() const
    {
        if (!head)
            throw runtime_error("empty");
        return head->val;
    }

   private:
    Node* head;
};

int main()
{
    LinkedListStack s;
    s.push(10);
    s.push(20);
    int v;
    if (s.pop(v))
        cout << v << "\n";
    cout << s.top() << "\n";
    return 0;
}

// Time Complexity: push/pop/top O(1)
// Space Complexity: O(n)
